### PR TITLE
SWIK-739 social login

### PR DESCRIPTION
**Pull request by TBoonX**
I added  new components and actions.
I enhanced the UserProfileStore, UserRegistrationStore, UserRegistration, user, userprofile
Also a fix for cookies in in there.

### CODE REVIEW

**Comment by kadevgraaf** [comment]
@rmeissn @TBoonX - Shall I or @bwulff QA this, for it to go in the update to stable?

**Comment by TBoonX** [comment]
Klaas its yours. But you have to test it with the user-service branch.
See https://slidewiki.atlassian.net/browse/SWIK-161

**Comment by kadevgraaf** [comment]
Hi Kurt!
Thanks. I talked with @bwulff  and we propose not to include the PR yet in the update to stable because it has a lot of new code/features on short notice/with little testing time. Would you agree or do you think it is quite safe and important to merge asap?

**Comment by TBoonX** [comment]
Let's discuss this in Genua -  perhaps the trial partner really need this. But yes we have to test it plus it increases the deployment difficulty level. Thus just add it to stable if really necessary

**Comment by abijames** [comment]
Also, as we are currently using the registration process to limit accounts to people associated with the project. we should only implement registration through social accounts when we have editing groups implemented.

**Comment by rmeissn** [comment]
I've tested social login. At all: Good work! 
Here's what I've noticed:

* Is it really necessary to tell people that a new tab will open up? Other WebApps just open a new tab (or better: a new and small window).
* I was confused by the dialog that told me that login via social provider failed. --> Of course it failed, because I had never logged in via this provider before. But the text should sound more like "Hey, you have never tried to login via Facebook. We will now create a new account for you at SlideWiki. Is that what you intended to do or do you want to try logging in through another provider?"
* Please add popups to the dialog where I had to add extra information that state why i cannot use my e.g. email address in case it is already used by another account.
* If possible, please also fetch country, language, bio and organization from the social provider and/or add the language of my browser as language
* The "Authorized Accounts" section in settings confused me, because it seemed like I already enabled Google and Github, but Facebook is missing. Actually, it is vice versa. Enabled providers should be green and not enabled ones should be grey.
* I should not be able to disable Facebook login in case this is my only login option --> please disable the "disable" button in case this is the only provider left.
* I have no possibility to delete the account at SlideWiki. At least, I should be able to delete/deactivate my account.
* I was not able to login again through facebook after I've logged out on slidewiki. SlideWiki told me that I have never logged in through facebook and need to create an account.

**Comment by TBoonX** [comment]
I thought it would be more accessible if there is a message. also it could take a few seconds until the opened tab closes itself - thus the user should be aware that he/she have to wait.
You are right, I have to remake the message for social login fail.
The popups of wrong fields are opened when the mouse is positioned over the input field. Do you want a error message box like in registration view?
biography is fetched but country is not possible beause I could get Germany, de, de_DE, Deutschland, .... I use for frontendLanguage always en_EN and for spokenLanguage the browser language.
OK I will change Integrations.js
Deactivate is now available. But we have to discuss how deactivated users have to be handled.
I couldn't validate your facebook failure. also with your credentials it just works.

**Comment by rmeissn** [comment]
"It could take a few seconds until the opened tab closes itself - thus the user should be aware that he/she have to wait."
--> I think this is normal for most (if not all) providers, except that they typically open a new (small) window instead of a tab.
"The popups of wrong fields are opened when the mouse is positioned over the input field. Do you want a error message box like in registration view?"
--> Seems like I just missed it....will test it again in the next test cycle
"biography is fetched but country is not possible beause I could get Germany, de, de_DE, Deutschland, .... I use for frontendLanguage always en_EN and for spokenLanguage the browser language."
--> okay
"Deactivate is now available. But we have to discuss how deactivated users have to be handled."
--> feel free to ;)
"I couldn't validate your facebook failure. also with your credentials it just works."
--> Will test again in the next test cycle

File: actions/user/deleteSocialData.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export default function deleteSocialData(context, payload, done) {
```
Comment (rmeissn) [inline actions/user/deleteSocialData.js]: You could add more than one method to an action (or file) by omitting the "default" and exporting several functions. Reimport is handled through `import {abc, xyz} from ...`

File: actions/user/registration/socialSignUp.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export default function socialSignUp(context, payload, done) {
+    context.service.create('user.socialregistration', payload, { timeout: 20 * 1000 }, (err, res) => {
+        if (err) {
+            context.dispatch('SOCIAL_CREATE_USER_FAILURE', err);
+            done();
```
Comment (rmeissn) [inline actions/user/registration/socialSignUp.js]: According to [Fluxible Action Docs](https://github.com/yahoo/fluxible/blob/master/packages/fluxible/docs/api/Actions.md) you could omit the `done()` call in case you're using promises. Also have a look at this [Fluxible Actions Docs](http://fluxible.io/api/actions.html) for correct usage of promises with dispatch.
The following should work:
```
function socialSignUp(context, payload) {
....
  return context.dispatch(....);
....
}
```

File: components/Login/LoginModal.js [hunk 3]
```diff
@@ -61,16 +67,61 @@ class LoginModal extends React.Component {
...
+        }
+    }
+
+    handleRegisterFirst(dismiss) {
+        if (dismiss === 'cancel')
+            return true;
+
+        localStorage.setItem(MODI, 'login_failed_register_now');
+
+        this.context.executeAction(deleteSocialData, { });
```
Comment (rmeissn) [inline components/Login/LoginModal.js]: Wait for the first action to finish? E.g. use of promises?

File: components/Login/LoginModal.js [hunk 4]
```diff
@@ -93,6 +144,124 @@ class LoginModal extends React.Component {
...
+
+            let win = window.open(url, '_blank', 'width='+width+',height='+height+',left='+left+',top='+topSpace+',toolbar=No,location=No,scrollbars=no,status=No,resizable=no,fullscreen=No');
+            win.focus();
+
+            return true;
+        })
+        .catch();
+    }
+
+    clickedFacebook(e) {
```
Comment (rmeissn) [inline components/Login/LoginModal.js]: Use a parameter to distinguish between different providers instead of implementing three functions with the (nearly) exact same content.

File: components/Login/LoginModal.js [hunk 5]
```diff
@@ -93,6 +144,124 @@ class LoginModal extends React.Component {
...
+            console.log('Error while parsing data', err);
+            return;
+        }
+        finally {
+            //delete data
+            localStorage.setItem(NAME, '');
+        }
+
+        //add language before send to service
+        let language = navigator.browserLanguage || navigator.language;
```
Comment (rmeissn) [inline components/Login/LoginModal.js]: In case this is used elsewhere too, add it as a common function. Seems to me like a common function.

File: components/Login/LoginModal.js [hunk 6]
```diff
@@ -93,6 +144,124 @@ class LoginModal extends React.Component {
...
+                confirmButtonText: 'Confirm',
+                confirmButtonClass: 'negative ui button',
+                buttonsStyling: false
+            }).then().catch();
+
+            return;
+        }
+
+        this.context.executeAction(newSocialData, data);
+        this.context.executeAction(userSocialSignIn, data);
```
Comment (rmeissn) [inline components/Login/LoginModal.js]: Wait for the first action to finish? E.g. use promises.

File: components/Login/LoginModal.js [hunk 7]
```diff
@@ -124,6 +293,12 @@ class LoginModal extends React.Component {
                         <div className="ui error message"/>
                       </form>
                       <br/>
+                      <div className="container">
+                        <i className="big circular facebook square link icon" onClick={this.clickedFacebook.bind(this)} ></i>
```
Comment (rmeissn) [inline components/Login/LoginModal.js]: See parameter hint from above

File: components/Login/Social.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+    findGetParameter(parameterName) {
+        let result = null,
+            index = 0;
+        location.search
+        .substr(1)
+            .split('&')
+            .forEach((item) => {
```
Comment (rmeissn) [inline components/Login/Social.js]: Use map+reduce if possible instead of a side effect (adding strings to result).

File: components/User/UserProfile/Integrations.js [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+                buttonsStyling: false
+            }).then(() => {
+                this.context.executeAction(resetProviderStuff, {});
+                return true;
+            }).catch();
+        }
+        else if (this.props.UserProfileStore.addProviderError === false && nextProps.UserProfileStore.addProviderError) {
+            swal({
+                title: 'Error',
+                text: 'The provider couldn&apos;t be added. Unknown error.',
```
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: Giving a user an "unkonwn error" is not helpful. It just confuses the user.

File: components/User/UserProfile/Integrations.js [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+        //prepare localStorage
+        localStorage.setItem(MODI, 'addProvider');
+        localStorage.setItem(NAME, '');
+
+        //observe storage
+        $(window).off('storage').on('storage', this.handleStorageEvent.bind(this));
+
+        //create new tab
+        let url = 'http://authorizationservice.manfredfris.ch:3000/connect/' + this.provider;
```
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: A hardcoded URL?!

File: components/User/UserProfile/Integrations.js [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+            this.context.executeAction(updateProviderAction, '');
+            return false;
+        }
+        finally {
+            //delete data
+            localStorage.setItem(NAME, '');
+        }
+
+        //add language before send to service
+        let language = navigator.browserLanguage || navigator.language;
```
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: The whole paragraph seems to be repeated in another file --> export into a function --> avoids copy and paste

File: components/User/UserProfile/Integrations.js [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+        if ( (data.token.length < 1)
+          || (data.provider.length < 3)
+          || (data.token_creation.length < 22) )
+            //Failure
+        {
+            this.context.executeAction(updateProviderAction, '');
+            return false;
+        }
+
+        if  (data.email.indexOf('@') === -1 || data.email.indexOf('.') === -1 || data.email.length < 5) {
```
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: A regex (search on stackoverflow) might be better to check for a valid email adress

File: components/User/UserProfile/Integrations.js [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+                        google = true;
+                        break;
+                    case 'github':
+                        github = true;
+                        break;
+                }
+            });
+        // console.log('Integrations render()', this.props.UserProfileStore.user.providers);
+
+        let facebook_icon_classes = classNames({
```
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: all of the following statements are either true or false. You might be able to condense rules to fewer lines by using `!facebook` or `facebook` more

File: components/User/UserProfile/Integrations.js [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+                    </p>
+                  </div>
+
+              </div>
+              <div className="ui segments">
+
+                  <div className="ui secondary segment">
+                      <h3>Login Provider</h3>
+                  </div>
+                  <div className="ui segment">
```
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: The following paragraphs have the exact same structure. Use a loop to cycle through the list of all providers (currently three) instead writing it three times.

File: components/User/UserProfile/UserProfile.js [hunk 15]
```diff
@@ -117,24 +120,31 @@ class UserProfile extends React.Component {
     }
 
     displayAccounts() {
-        return (<div>
-            <div className="ui segments">
-            <div className="ui secondary segment">
-              <h3>Change password</h3>
-            </div>
+        let changePassword = '';
```
Comment (rmeissn) [inline components/User/UserProfile/UserProfile.js]: Use an inline if else statement .... `let tmp = (condition) ? 'yes' : 'no' `

File: components/User/UserProfile/UserProfile.js [hunk 16]
```diff
@@ -143,8 +153,14 @@ class UserProfile extends React.Component {
         return (<PrivatePublicUserProfile user={this.props.UserProfileStore.user} decks={this.props.UserProfileStore.userDecks}/>);
     }
 
+    displayIntegrations() {
```
Comment (rmeissn) [inline components/User/UserProfile/UserProfile.js]: If it's that easy, you might be able to omit the function at all by moving `<Integrations />` upwards.

File: components/User/UserRegistration/UserRegistration.js [hunk 17]
```diff
@@ -116,10 +134,55 @@ class UserRegistration extends React.Component {
             return (usernameNotAllowed !== undefined) ? !usernameNotAllowed : true;
         });
 
-        $('.ui.form').form(validationRules);
+        $('.ui.form').not('.registrationmodalform').form(validationRules);
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Use react identifiers instead of searching for classes! Searching for classes caused problems before.

File: components/User/UserRegistration/UserRegistration.js [hunk 18]
```diff
@@ -116,10 +134,55 @@ class UserRegistration extends React.Component {
...
+            })
+            .then((dismiss) => {
+                if (dismiss === 'cancel')
+                    return true;
+
+                this.context.executeAction(navigateAction, {
+                    url: '/'
+                });
+
+                $('.ui.login.modal').modal('show');
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Use react identifiers.

File: components/User/UserRegistration/UserRegistration.js [hunk 19]
```diff
@@ -116,10 +134,55 @@ class UserRegistration extends React.Component {
...
+
+                return true;
+            })
+            .catch(() => {
+                $('.ui.socialregistration.modal').modal('hide');
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Use react identifiers.

File: components/User/UserRegistration/UserRegistration.js [hunk 20]
```diff
@@ -170,21 +233,32 @@ class UserRegistration extends React.Component {
...
+                firstname: this.refs.firstname.value,
+                lastname: this.refs.lastname.value,
+                username: this.refs.username.value,
+                language: language,
+                email: this.refs.email.value,
+                password: hashPassword(this.refs.password.value),
+                grecaptcharesponse: this.state.grecaptcharesponse
+            };
+        } catch (e) {
+            //somehow this is sometimes called when doing login
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Please debug it instead of having a try catch statement

File: components/User/UserRegistration/UserRegistration.js [hunk 21]
```diff
@@ -208,6 +282,160 @@ class UserRegistration extends React.Component {
...
+        localStorage.setItem(NAME, '');
+
+        this.provider = provider;
+
+        //observe storage
+        $(window).off('storage').on('storage', this.handleStorageEvent.bind(this));
+
+        //show hint before open tab
+        const Provider = this.getProviderName();
+        swal({
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Maybe not needed at all? Just open the login window.

File: components/User/UserRegistration/UserRegistration.js [hunk 22]
```diff
@@ -208,6 +282,160 @@ class UserRegistration extends React.Component {
...
+
+            let win = window.open(url, '_blank', 'width='+width+',height='+height+',left='+left+',top='+topSpace+',toolbar=No,location=No,scrollbars=no,status=No,resizable=no,fullscreen=No');
+            win.focus();
+
+            return true;
+        })
+        .catch();
+    }
+
+    clickedFacebook(e) {
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Copied and pasted from above ... export to an external function/file

File: components/User/UserRegistration/UserRegistration.js [hunk 23]
```diff
@@ -296,9 +525,23 @@ class UserRegistration extends React.Component {
                         <div >
                             By clicking Sign Up, you agree to our <a href="">Terms</a>.
                         </div>
+                        <div className="ui dividing header" ></div>
+                        <div className="container">
+                            Register with another platform:
+                            <br/>
+                            <i className="big circular facebook square link icon" onClick={this.clickedFacebook.bind(this)} ></i>
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: User parameters (like already written above)

File: components/User/UserRegistration/UserRegistrationSocial.js [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+import ReactDOM from 'react-dom';
+import classNames from 'classnames';
+import {navigateAction} from 'fluxible-router';
+import {connectToStores} from 'fluxible-addons-react';
+import socialSignUp from '../../../actions/user/registration/socialSignUp';
+import checkEmail from '../../../actions/user/registration/checkEmail';
+import checkUsername from '../../../actions/user/registration/checkUsername';
+import UserRegistrationStore from '../../../stores/UserRegistrationStore';
+
+const headerStyle = {
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistrationSocial.js]: I saw this before ... copy and pasted. Avoid it by exporting it.

File: components/User/UserRegistration/UserRegistrationSocial.js [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+        e.preventDefault();
+
+        $('.ui.socialregistration.modal').modal('hide');
+        let user = this.props.UserRegistrationStore.socialuserdata;
+        user.email = this.refs.email.value;
+        user.username = this.refs.username.value;
+        user.forename = this.refs.firstname.value;
+        user.surname = this.refs.lastname.value;
+
+        let language = navigator.browserLanguage ? navigator.browserLanguage : navigator.language;
```
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistrationSocial.js]: Copied and pasted from above ... export to an external function/file

File: services/user.js [hunk 26]
```diff
@@ -44,6 +44,31 @@ export default {
                         error: err
                     });
                 });
+        } else if (resource === 'user.socialsignin') {
+            rp.post({
+                uri: Microservices.user.uri + '/social/login',
+                body: JSON.stringify({
```
Comment (rmeissn) [inline services/user.js]: Just stringify args instead?!

File: stores/UserProfileStore.js [hunk 27]
```diff
@@ -32,6 +32,11 @@ class UserProfileStore extends BaseStore {
         this.jwt = '';
         this.userpicture = undefined;
         this.errorMessage = '';
+        this.socialLoginError = false;
```
Comment (rmeissn) [inline stores/UserProfileStore.js]: use an object instead of several independent items

File: stores/UserProfileStore.js [hunk 28]
```diff
@@ -198,6 +218,11 @@ class UserProfileStore extends BaseStore {
         this.emitChange();
     }
 
+    handleSocialSignInError(err) {
+        this.socialLoginError = true;
```
Comment (rmeissn) [inline stores/UserProfileStore.js]: You might be able to reset it immediately after `this.emitChange()` instead of having a full component --> action --> store cycle again.

File: stores/UserProfileStore.js [hunk 29]
```diff
@@ -207,6 +232,68 @@ class UserProfileStore extends BaseStore {
...
         return message2;
     }
+
+    deleteSocialData() {
+        this.socialLoginError = false;
+        this.emitChange();
+    }
+
+    socialRegister(res) {
+        this.userpicture = res.picture;
```
Comment (rmeissn) [inline stores/UserProfileStore.js]: Why is this needed?

File: components/User/UserProfile/Integrations.js (hunk 13 above)
Comment (TBoonX) [inline components/User/UserProfile/Integrations.js]: As I have three flags its better for readers to not change it

File: components/User/UserProfile/UserProfile.js (hunk 16 above)
Comment (TBoonX) [inline components/User/UserProfile/UserProfile.js]: I will not do this, because this will break the used style

File: services/user.js (hunk 26 above)
Comment (TBoonX) [inline services/user.js]: I tried it but the service acted very strange. As the other calls to this service also use stringify I will not change it. Also the parameter "json: true" just do the stringify for you, thus no overhead.

File: stores/UserProfileStore.js (hunk 28 above)
Comment (TBoonX) [inline stores/UserProfileStore.js]: Done.

File: components/User/UserRegistration/UserRegistration.js (hunk 18 above)
Comment (rmeissn) [inline components/User/UserRegistration/UserRegistration.js]: Hm, good question. It should be.....try to search in the React documentation.

File: components/User/UserProfile/Integrations.js (hunk 13 above)
Comment (rmeissn) [inline components/User/UserProfile/Integrations.js]: Depends on your attitude ;) For me, it's easier to read in parameters and I'm confused by repeated functions that do the same thing :P

**Comment by rmeissn** [comment]
Executed a functional test together with Kurt. We found some bugs that need to be fixed before it may be merged.

**Comment by rmeissn** [comment]
Executed server side tests, everything's fine, ready to merge
